File: src/main/kotlin/fr/jhelp/kotlinLightSamples/blueetooth/BluetoothFrame.kt
package fr.jhelp.kotlinLightSamples.blueetooth

import fr.jhelp.kotlinLight.ByteStream
import fr.jhelp.kotlinLight.CommonList

class BluetoothFrame
{
    companion object
    {
        fun mergeUInt16(highPart: Int, lowPart: Int): Int
        {
            return ((highPart and 0xFFFF) shl 16) or (lowPart and 0xFFFF)
        }

        fun getBits(input: Int, offset: Int, length: Int): Int
        {
            return (input shr offset) and ((1 shl length) - 1)
        }
    }

    private val byteStream: ByteStream

    constructor(byteArray: CommonList<Byte>)
    {
        this.byteStream = ByteStream(byteArray, false)
    }

    fun getTimestamp(): Long
    {
        return this.byteStream.readInt32().toLong() and 0xFFFFFFFFL
    }

    fun getUInt8(): Int
    {
        return this.byteStream.readInt8()
    }

    fun getUInt16(): Int
    {
        return this.byteStream.readInt16()
    }

    fun getUInt32(): Long
    {
        return this.byteStream.readInt32().toLong() and 0xFFFFFFFFL
    }

    fun getInt8(): Int
    {
        return this.byteStream.readInt8().toByte().toInt()
    }

    fun getInt16(): Int
    {
        var value = this.byteStream.readInt16()

        if ( (value and 0x8000) != 0)
        {
            value -= 0x10000
        }

        return value
    }

    fun getInt32(): Int
    {
        return this.byteStream.readInt32()
    }

    fun getFloat(): Float
    {
        return this.byteStream.readFloat()
    }

    fun position(position: Int): BluetoothFrame
    {
        this.byteStream.setPosition(position)
        return this
    }

    fun remaining(): Int
    {
        return this.byteStream.remaining()
    }

    fun rewind(): BluetoothFrame
    {
        this.byteStream.rewind()
        return this
    }
}

File: src/main/swift/Tests/swiftTests/src/jhelp/kotlinLightSamples/blueetooth/BluetoothFrame.swift


public class BluetoothFrame {

    
        public static func mergeUInt16(_ highPart: Int, _ lowPart: Int) -> Int
        {
            return ((highPart & 0xFFFF) << 16) | (lowPart & 0xFFFF)
        }

        public static func getBits(_ input: Int, _ offset: Int, _ length: Int) -> Int
        {
            return (input >> offset) & ((1 << length) - 1)
        }
    

    private let byteStream: ByteStream

    public init(_ byteArray: Array<Byte>){
        self.byteStream = ByteStream(byteArray, false)
    }

    public func getTimestamp() -> Long
    {
        return self.byteStream.readInt32().toLong() & Long(0xFFFFFFFF)
    }

    public func getUInt8() -> Int
    {
        return self.byteStream.readInt8()
    }

    public func getUInt16() -> Int
    {
        return self.byteStream.readInt16()
    }

    public func getUInt32() -> Long
    {
        return self.byteStream.readInt32().toLong() & Long(0xFFFFFFFF)
    }

    public func getInt8() -> Int
    {
        return self.byteStream.readInt8().toByte().toInt()
    }

    public func getInt16() -> Int
    {
        var value = self.byteStream.readInt16()

        if  (value & 0x8000) != 0
        {
            value -= 0x10000
        }

        return value
    }

    public func getInt32() -> Int
    {
        return self.byteStream.readInt32()
    }

    public func getFloat() -> Float
    {
        return self.byteStream.readFloat()
    }

    public func position(_ position: Int) -> BluetoothFrame
    {
        self.byteStream.setPosition(position)
        return self
    }

    public func remaining() -> Int
    {
        return self.byteStream.remaining()
    }

    public func rewind() -> BluetoothFrame
    {
        self.byteStream.rewind()
        return self
    }
}
